evaluate.go

```diff
@@ -142,186 +141,133 @@ func doMatchIsEmpty(matcher *MatchExpression, value reflect.Value) (bool, error)
-	//       2. If expression.Value could be converted it will already have been. No need to try
-	//          and convert again. There is also no need to check that the types match as they MUST
-	//          in order to have passed validation.
-	//       3. If we are presented with a map and we have more selectors to go through then its key
-	//          type MUST be a string
-	//       4. We already have validated that the operations can be performed on the target data.
-	//          So calls to the doMatch* functions don't need to do any checking to ensure that
-	//          calling various fns on them will work and not panic - because they wont.
-
-	if depth >= len(expression.Selector.Path) {
-		// we have reached the end of the selector - execute the match operations
-		switch expression.Operator {
-		case MatchEqual:
-			return doMatchEqual(expression, rvalue)
-		case MatchNotEqual:
-			result, err := doMatchEqual(expression, rvalue)
-			if err == nil {
-				return !result, nil
-			}
-			return false, err
-		case MatchIn:
-			return doMatchIn(expression, rvalue)
-		case MatchNotIn:
-			result, err := doMatchIn(expression, rvalue)
-			if err == nil {
-				return !result, nil
-			}
-			return false, err
-		case MatchIsEmpty:
-			return doMatchIsEmpty(expression, rvalue)
-		case MatchIsNotEmpty:
-			result, err := doMatchIsEmpty(expression, rvalue)
-			if err == nil {
-				return !result, nil
-			}
-			return false, err
-		case MatchMatches:
-			return doMatchMatches(expression, rvalue)
-		case MatchNotMatches:
-			result, err := doMatchMatches(expression, rvalue)
-			if err == nil {
-				return !result, nil
-			}
-			return false, err
-		default:
-			return false, fmt.Errorf("Invalid match operation: %d", expression.Operator)
-		}
-	}
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		return CoerceUint64(expression.Value.Raw)
 
-	switch rvalue.Kind() {
-	case reflect.Struct:
-		fieldName := expression.Selector.Path[depth]
-		fieldConfig := fields[FieldName(fieldName)]
+	case reflect.Float32:
+		return CoerceFloat32(expression.Value.Raw)
 
-		if fieldConfig.StructFieldName != "" {
-			fieldName = fieldConfig.StructFieldName
-		}
+	case reflect.Float64:
+		return CoerceFloat64(expression.Value.Raw)
 
-		value := reflect.Indirect(rvalue.FieldByName(fieldName))
+	default:
+		return expression.Value.Raw, nil
+	}
+}
 
-		if matcher, ok := value.Interface().(MatchExpressionEvaluator); ok {
-			return matcher.EvaluateMatch(Selector{
-				Type: expression.Selector.Type,
-				Path: expression.Selector.Path[depth+1:],
-			}, expression.Operator, getMatchExprValue(expression))
-		}
+func evaluateMatchExpression(expression *MatchExpression, datum interface{}) (bool, error) {
+	path := fmt.Sprintf("/%s", strings.Join(expression.Selector.Path, "/"))
+	ptr, err := pointerstructure.Parse(path)
+	if err != nil {
+		return false, fmt.Errorf("error parsing path: %w", err)
+	}
+	ptr.Config.TagName = "bexpr"
```

Comment:
Looking back at the previous PR I see that we are not doing any of the JSON pointer ~ expansion in the parser. Which means that we somehow need to do it and going through pointerstructure.Parse here is probably what is making that happen.

I wonder whether it wouldn't be better to do the expansion in the parser and then have the `Selector.String` method gain some smarts to convert `/` -> `~1` and `~` -> `~0` within each part of the path prior to outputting.

So I guess that is a potential optimization and as long as the benchmarks still seem reasonable its probably not something we need to care about right now.

Reply:
Pointerstructure does that expansion -- why do you think we need to do that in this library? I don't see any reason to do that expansion, then have to undo it before passing it to pointerstructure.

Reply:
`pointerstructure` only does that expansion in the `Parse` method. My thinking is that there is no need to ever call `Parse` as we already have the path parts split out and can just create a new `Pointer` on the stack. Not calling `Parse` would require us to do the expansion somewhere but would save us:

* 1 string formatting which will result in at least 1 memory allocation
* 1 `pointerstructure.Parse` call which will result in:
   * 1 allocation of a `Pointer` type
   * 1+ allocations for the slice to hold the pointer path
   * Extra processing that we already did in the `bexpr` parser

So while this code functions perfectly fine, there are optimizations that could be made to perform slightly better. Since it probably wont be an issue the code here is fine and we can optimize later if we find out its a bottleneck.